### PR TITLE
Document the new `buildable` indirect selection mode

### LINKED ISSUE #2568: `buildable` indirect selection mode
### Contributions

- [X] I am a PM or subject matter expert at dbt who is responsible for this feature.

### Where does this content belong?

https://docs.getdbt.com/reference/node-selection/test-selection-examples#indirect-selection

### Links
- https://github.com/dbt-labs/dbt-core/issues/6365
- https://github.com/dbt-labs/dbt-core/pull/6366

### Who will use this feature?
The intended audience is an analytics engineer iteratively developing a subset of the DAG, likely modifying pre-existing models.

### Feature description how it will help people
"This will benefit developers who are testing changes on a subset of the graph, rather than building the entire graph."
The author's organization typically has tests that compare models with their dependencies. e.g. confirm that the transactions add up to the same amount -- if they differ, that would be bad! During development, the'd want to run the minimal number of nodes to check that things are working rather than running the entire DAG.

### Customer feedback and research
Community member @AGPapa recognized the need and opened a pull request with a solution 💪 @MichelleArk has dealt with similarly complex and specialized selection criteria with a large DAG shared over multiple teams.

### How can we test or use this feature?
1. Create model `model_a` (probably as a `seed` file)
1. Create model `model_b` that depends only on `model_a`
1. Create model `model_c` that depends only on `model_b`
1. Create test `test_ab` that depends on `model_a` and `model_b`
1. Create test `test_cb` that depends on `model_c` and `model_b`

To compare the 3 modes:
```shell
dbt test --select model_b --indirect-selection=buildable
dbt test --select model_b --indirect-selection=cautious
dbt test --select model_b --indirect-selection=eager
```

### Screenshot
Example of which modes will select each multi-parent test:

image

### 🚧 How could the existing page be improved?

A few things within [this](https://docs.getdbt.com/reference/node-selection/test-selection-examples#indirect-selection) section:
1. There is no stand-alone example for `indirect_selection: cautious` in a [yaml selector](https://docs.getdbt.com/reference/node-selection/yaml-selectors)
1. The [indirect selection](https://docs.getdbt.com/reference/node-selection/test-selection-examples#indirect-selection) section is verbose, but some of us aren't good at reading long text (this guy!) 😬 
1. `$ dbt test --select customers orders` is included, but not explained

### PR DESCRIPTION
**Pull request by dbeatty10**
resolves #2568

## Preview Links

- https://deploy-preview-2858--docs-getdbt-com.netlify.app/reference/node-selection/test-selection-examples
- https://deploy-preview-2858--docs-getdbt-com.netlify.app/reference/node-selection/yaml-selectors#indirect-selection

## Design aspirations
1. Document the new `buildable` mode
1. Include examples using `dbt build` (since those are missing currently)
1. Avoid adding much more length to the existing docs
1. Be as understandable as possible
1. Include versioning components since `buildable` is not available until dbt-core 1.4

### Comparison of lengths

7.5% more words were added in 046fbb771a627a5367b7ba4d8644cff3f06704d0.

|        | Lines | Words | Bytes |
|--------|-------|-------|-------|
| Before | 195   | 1056  | 7215  |
| After  | 221   | 1136  | 7700  |

Before:
```
wc website/docs/reference/node-selection/test-selection-examples.md                                                 
     195    1056    7215 website/docs/reference/node-selection/test-selection-examples.md
```

After:
```
wc website/docs/reference/node-selection/test-selection-examples.md                                                 
     221    1136    7700 website/docs/reference/node-selection/test-selection-examples.md
```

#### Test selection examples

image

#### YAML selectors

image

image

#### Test selection examples

image

#### YAML selectors

image

image

## Checklist

- [x] Add versioning components, as described in [Versioning Docs](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/versioningdocs.md)
- [x] Review the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) and [About versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) so my content adheres to these guidelines.

### CODE REVIEW

File: website/docs/reference/node-selection/test-selection-examples.md [hunk 1]
```diff
@@ -45,32 +45,80 @@ In both cases, `test_type` checks a property of the test itself. These are forms
 
 ### Indirect selection
 
+<VersionBlock lastVersion="1.3">
 
-  ```bash
-  $ dbt test --select customers
-  $ dbt test --select orders
-  ```
+#### Two modes for indirect selection
```
Comment (mirnawong1) [inline website/docs/reference/node-selection/test-selection-examples.md]: sugg removing this sub header and enveloping in the sentence

```suggestion

```

File: website/docs/reference/node-selection/test-selection-examples.md [hunk 2]
```diff
@@ -45,32 +45,80 @@ In both cases, `test_type` checks a property of the test itself. These are forms
...
+There are two options to configure the behavior when performing indirect selection (with `eager` as the default):
```
Comment (mirnawong1) [inline website/docs/reference/node-selection/test-selection-examples.md]: suggesting this IF it's important to use the word 'mode'

```suggestion
There are two modes (or options) to configure the behavior when performing indirect selection (with `eager` as the default):
```

File: website/docs/reference/node-selection/test-selection-examples.md [hunk 3]
```diff
@@ -45,32 +45,80 @@ In both cases, `test_type` checks a property of the test itself. These are forms
...
+Note that test exclusion is always greedy: if ANY parent is explicitly excluded, the test will be excluded as well.
+
+The "cautious" mode can be useful in environments when you're only building a subset of your DAG, and you want to avoid test failures in "eager" mode caused by unbuilt resources. (Another way to achieve this is with [deferral](defer)).
+
+</VersionBlock>
+
+<VersionBlock firstVersion="1.4">
 
-These are examples of "indirect" selection: `customers` and `orders` select models (whether by name or path). Any tests defined on either `customers` or `orders` will be selected indirectly, and thereby included.
+#### Three modes for indirect selection
```
Comment (mirnawong1) [inline website/docs/reference/node-selection/test-selection-examples.md]:
```suggestion

```

File: website/docs/reference/node-selection/test-selection-examples.md [hunk 4]
```diff
@@ -45,32 +45,80 @@ In both cases, `test_type` checks a property of the test itself. These are forms
...
+There are three options to configure the behavior when performing indirect selection (with `eager` as the default):
```
Comment (mirnawong1) [inline website/docs/reference/node-selection/test-selection-examples.md]: suggesting this IF it's important to use the word 'mode'

```suggestion
There are three modes (or options) to configure the behavior when performing indirect selection (with `eager` as the default):
```

File: website/docs/reference/node-selection/yaml-selectors.md [hunk 5]
```diff
@@ -112,6 +140,8 @@ where we can only pass one "yeslist" (`--select`) and one "nolist" (`--exclude`)
 
 #### Indirect selection
```
Comment (mirnawong1) [inline website/docs/reference/node-selection/yaml-selectors.md:141]: i know this already exists but do you think there's value in making these (+ the exclude h4 header) into H3 headers?

**Comment by mirnawong1** [comment]
hey @dbeatty10 this is awesome! thank you very much for super detailed explanation - really helps!

left some small tweaks. do you think it'll be a better ux if we create tabs for the eager, cautious, and buildable modes? i took the liberty of creating tabs code for the particular version - here's a visual, what do you think? 

Screenshot 2023-02-14 at 13 33 07


for some reason i can't upload the code format correctly so I'll dm it to you and also paste it in [this notion task](https://www.notion.so/dbtlabs/Review-Doug-s-PR-and-provide-tabs-code-076fc6453c314446876aaba793e19ff7?pvs=4) for you to view.

File: website/docs/reference/node-selection/test-selection-examples.md [hunk 6]
```diff
@@ -45,32 +45,80 @@ In both cases, `test_type` checks a property of the test itself. These are forms
...
+```shell
+$ dbt test --select orders --indirect-selection=buildable
+$ dbt build --select orders --indirect-selection=buildable
+```
+
+This mode is similarly conservative like "cautious", but is slightly more inclusive.
+
+It will only include tests whose references are each within the selected nodes (or their ancestors).
+
+This is useful in the same scenariors as "cautious", but also includes when a test depends on a model **and** a direct ancestor of that model (like confirming an aggregation has the same totals as its input).
```
Comment (mirnawong1) [inline website/docs/reference/node-selection/test-selection-examples.md]: typo and also corrected in the tabs recommendation

```suggestion
This is useful in the same scenarios as "cautious", but also includes when a test depends on a model **and** a direct ancestor of that model (like confirming an aggregation has the same totals as its input).
```

File: website/docs/reference/node-selection/yaml-selectors.md (hunk 5 above)
Comment (dbeatty10) [reply website/docs/reference/node-selection/yaml-selectors.md:141]: Yeah, would be great to take a look at upgrading the headers for the whole page.

**Comment by dbeatty10** [comment]
> do you think it'll be a better ux if we create tabs for the eager, cautious, and buildable modes? i took the liberty of creating tabs code for the particular version - here's a visual, what do you think?

I love the way that the tabs turned out!

The one thing I'm curious about -- is there a way to create link that will go directly to the relevant tab?

When I'm having a discussion, I frequently will share a link that goes _exactly_ to the relevant sub-section of the docs, like this:
- https://docs.getdbt.com/reference/node-selection/test-selection-examples#run-tagged-tests-only

If we can do that with tabs, that would be awesome.

I'll try pushing a commit with your version of the Markdown that re-formats using tabs so we can play around with it.

**Comment by dbeatty10** [comment]
### Using H4 headers

image

### Using tabs

I haven't moved the code example to the end yet. Wanted to keep the diffs small first while we work this out. Will plan to move them down there though.

image

**Comment by mirnawong1** [comment]
> 

that's a great question - let me ask the dev front end team! i wonder if adding an anchor tag will do the trick! or maybe docusaurus already has this https://docusaurus.io/docs/next/markdown-features/tabs

update: I've asked in [this slack thread](https://dbt-labs.slack.com/archives/C01C2H652BC/p1676395378034749)

**Comment by dbeatty10** [comment]
> > 
> 
> that's a great question - let me ask the dev front end team! i wonder if adding an anchor tag will do the trick! or maybe docusaurus already has this https://docusaurus.io/docs/next/markdown-features/tabs
> 
> update: I've asked in [this slack thread](https://dbt-labs.slack.com/archives/C01C2H652BC/p1676395378034749)

I've added a [query string](https://docusaurus.io/docs/next/markdown-features/tabs?current-os=android#query-string)  (`queryString` prop) within fcb85b2b388f2d648169662c2ffd8b9da3b7e7cc -- we'll see how it works!

**Comment by dbeatty10** [comment]
> I've added a [query string](https://docusaurus.io/docs/next/markdown-features/tabs?current-os=android#query-string) (`queryString` prop) within [fcb85b2](https://github.com/dbt-labs/docs.getdbt.com/commit/fcb85b2b388f2d648169662c2ffd8b9da3b7e7cc) -- we'll see how it works!

At least with the Netlify preview, the `queryString` prop didn't work for me ☹️ -- not sure if I did something incorrectly:
- https://deploy-preview-2858--docs-getdbt-com.netlify.app/reference/node-selection/test-selection-examples?indirect-selection-mode=buildable

**Comment by dbeatty10** [comment]
@mirnawong1 I think I've followed up on all your (amazing!) suggestions.

One thing I did change:
- it felt like we should standardize on either "modes" or "options", and I picked the former since it was pre-existing

Open to standardizing on "options" instead. Or if you think it best, we could go back to "modes (or options)" or "options (or modes)".

So at least two outstanding things to make a decision on before this PR is done:
- "modes" vs. "options" vs. both
- determine if we need/want [query string](https://docusaurus.io/docs/next/markdown-features/tabs?current-os=android#query-string) (`queryString` prop) or not. Personally, I want it, but I don't know how long I'm willing to spend troubleshooting it 😅

**Comment by mirnawong1** [comment]
Hey @dbeatty10 sounds awesome! I'm more than happy to go with your gut with regards to options or modes. I think modes makes sense as well. 

In terms of the tabs and query string, I don't think this is going to be done asap as it requires the front end team to upgrade docusaurus. If we want to merge this pr without it, it would mean either:
- using tabs but linking out to the "indirect selection" sub header since it doesn't work in tabs just yet. OR
- not using tabs and keeping it with sub headers.

I will absolutely look at this in more detail but I wanted to address your comment now in preparation for tomorrow 🤩🙏

**Comment by dbeatty10** [comment]
Sounds good @mirnawong1 !

We can definitely pick this up again tomorrow sometime.

I'd feel good about:
- using tabs but linking out to the "indirect selection" sub header

That way we can merge now and update later once Docusaurus is upgraded.

I'm ready for merge whenever you are. Also willing to dial in any other details you see that are outstanding.

**Review by mirnawong1 — APPROVED** [review]
thanks @dbeatty10 and deploy previews looks awesome! approved for you to merge ✨

**Comment by dbeatty10** [comment]
Thank you so much for all your help @mirnawong1 ! Just enabled auto-merge (so that it will automatically squash and merge once all requirements are met).

Opened this issue to track the future work for links for tabbed content:
- https://github.com/dbt-labs/docs.getdbt.com/issues/2874

**Comment by mirnawong1** [comment]
that's awesome, thank you so much @dbeatty10 - you beat me to it! i will comment in the asana task as well so I'm notified about the docusaurus upgrade

**Pull request merged**